File: client/lib/student-service.ts
import { authFetch } from "@/lib/api"
import { apiUrl } from "@/lib/config"
import { Students } from "@/app/(default)/students/columns"

const BASE_URL = `${apiUrl("/api/students")}/`

export async function getStudents(): Promise<Students[]> {
  const res = await fetch(`${BASE_URL}`)
  if (!res.ok) {
    const message = await res.text().catch(() => "")
    throw new Error(message || "Failed to fetch students")
  }
  return res.json()
}

export async function addStudent(values: {
  idNo: string
  firstName: string
  lastName: string
  course: string
  year: string
  gender: string
  photo_path?: string
}) {
  const payload = { ...values, year: parseInt(values.year) }

  const res = await authFetch(`${BASE_URL}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify(payload),
  })

  if (!res.ok) {
    const message = await res.text().catch(() => "")
    throw new Error(message || "Failed to add student")
  }
}

export async function editStudent(
  oldIdNo: string,
  values: {
    idNo: string
    firstName: string
    lastName: string
    course: string
    year: string
    gender: string
    photo_path?: string
  }
) {
  const payload = { ...values, year: parseInt(values.year) }

  const res = await authFetch(`${BASE_URL}${oldIdNo}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify(payload),
  })

  if (!res.ok) {
    const message = await res.text().catch(() => "")
    throw new Error(message || "Failed to update student")
  }
}

export async function deleteStudent(idNo: string) {
  const res = await authFetch(`${BASE_URL}${idNo}`, {
    method: "DELETE",
    credentials: "include",
  })

  if (!res.ok) {
    const message = await res.text().catch(() => "")
    throw new Error(message || "Failed to delete student")
  }
}

export async function bulkDeleteStudents(ids: string[]) {
  const res = await authFetch(`${apiUrl("/api/students/bulk-delete")}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ids }),
  })

  if (!res.ok) {
    const message = await res.text().catch(() => "")
    throw new Error(message || "Failed to bulk delete students")
  }
}

File: server/services/business/student_service.py
from repositories import student_repository


def _format_student(row):
    """Format a student row from the database."""
    return {
        "idNo": row["idNo"],
        "firstName": row["firstName"],
        "lastName": row["lastName"],
        "college_code": row.get("college_code"),
        "course": row.get("course"),
        "year": row.get("year"),
        "gender": row.get("gender"),
        "photo_path": row.get("photo_path"),
    }


def get_all():
    """Get all students."""
    rows = student_repository.get_all()
    return [_format_student(r) for r in rows]


def get_by_id(id_no):
    """Get a single student by idNo."""
    row = student_repository.get_by_id(id_no)
    return _format_student(row) if row else None


def create(id_no, first_name, last_name, course, year, gender, photo_path):
    """Create a new student."""
    row = student_repository.create(
        id_no, first_name, last_name, course, year, gender, photo_path
    )
    return _format_student(row)


def update(id_no, new_id_no, first_name, last_name, course, year, gender, photo_path):
    """Update a student by idNo."""
    row = student_repository.update(
        id_no, new_id_no, first_name, last_name, course, year, gender, photo_path
    )
    return _format_student(row) if row else None


def delete(id_no):
    """Delete a student by idNo."""
    row = student_repository.delete(id_no)
    return _format_student(row) if row else None


def bulk_delete(ids):
    """Delete multiple students by idNo."""
    deleted_rows = student_repository.bulk_delete(ids)
    return len(deleted_rows)
